src/uct/ib/ud/base/ud_iface.h

```diff
@@ -250,10 +255,34 @@ static UCS_F_ALWAYS_INLINE void uct_ud_leave(uct_ud_iface_t *iface)
     UCS_ASYNC_UNBLOCK(iface->super.super.worker->async);
 }
 
+static UCS_F_ALWAYS_INLINE void
+uct_ud_grh_get_dgid(const struct ibv_grh *grh, union ibv_gid *dgid)
+{
+    uint8_t ipmask = 0xf0;
+    uint8_t ipver  = ((*(uint8_t*)grh) & ipmask) >> 4;
+    size_t i;
+
+    if (ipver == 6) {
+        *dgid = grh->dgid;
+    } else {
+        for (i = 0; i < (sizeof(union ibv_gid) - UCT_UD_IPV4_ADDR_LEN - 2); ) {
+            dgid->raw[i++] = 0x00;
+        }
+
+        dgid->raw[i++]     = 0xff;
+        dgid->raw[i++]     = 0xff;
+
+        memcpy(&dgid->raw[i], &grh->dgid.raw[i], UCT_UD_IPV4_ADDR_LEN);
```

Comment:
do you really need to do a copy? Why not return a pointer?

Reply:
dgid with IPV4_ADDR_LEN placed in GRH can start with uninitialized values

Reply:
or you mean we can re-write GRH inplace?

Reply:
yes, probably rewrite option is worth checking, if this filling is really needed

Reply:
done